mlir/compiler-course/dudchenko_o_mlir/dudchenko_o_mlir.cpp

```diff
@@ -0,0 +1,74 @@
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Tools/Plugins/PassPlugin.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace mlir;
+
+namespace {
+
+void replaceCeil(ModuleOp module) {
+
+  for (auto func : module.getOps<mlir::func::FuncOp>()) {
+
+    func.walk([&](Operation *op) {
+      if (auto ceilOp = dyn_cast<math::CeilOp>(op)) {
+        Location loc = ceilOp->getLoc();
+        Value input = ceilOp->getOperand(0);
+        auto type = input.getType();
+
+        if (auto vectorType = type.dyn_cast<VectorType>()) {
+          if (!vectorType.getElementType().isa<FloatType>())
+            return;
+        } else if (!type.isa<FloatType>()) {
```

Comment:
Prefer separate functions instead of built-in methods.
```suggestion
        } else if (!isa<FloatType>(type)) {
```